### ISSUE TITLE
`doc` file can be submitted as a code

### ISSUE THREAD

**Issue by HeRaNO** [DOMjudge/domjudge#2218]
### Description of the problem

`doc` file can be submitted as code. It seems the file bypasses the file extension check.

### Your environment

- DOMjudge 8.2.2
- Ubuntu 20.04
- Nginx server

### Steps to reproduce

[A.zip](https://github.com/DOMjudge/domjudge/files/13400004/A.zip)

Unzip the archive first, and submit the `A.doc`  to any problem. Language setting sets as default.

![setting](https://github.com/DOMjudge/domjudge/assets/18643142/769f3ee4-932e-40f2-be56-c9575d6baf7b)

### Expected behaviour

![expect](https://github.com/DOMjudge/domjudge/assets/18643142/36ebee31-deec-4880-b8d5-ad40a3efdd7a)

### Actual behaviour

![actual](https://github.com/DOMjudge/domjudge/assets/18643142/d0e6f10a-7786-4291-b5e2-ffaa1b319718)

![backend](https://github.com/DOMjudge/domjudge/assets/18643142/386e2297-f2f6-4828-b18b-6da6922c0a3b)

**Comment by vmcj** [comment]
Hi @HeRaNO, I've tried to reproduce this and I can't. Can you maybe upload a bit more screenshots on how you submit this?

When I upload a file ending with the `.doc` syntax via submit the following field will stay empty: `Language` but for you this got picked up as `C`? Or did you manually select `C` as language?

If it's the first case this is indeed a bug, in the second case the message `match any of the allowed extensions` would be wrong as a participant can decide to hand something in with a custom extension (think about someone who can decide to use `.py2` vs `.py3`) and the system should just leave the language selection to the team IMO.

**Comment by ItsNiklas** [comment]
I think the problem is that when the option `Filter files passed to compiler by extension list` is enabled, the system only [checks if the filename ends with 'c'](https://github.com/DOMjudge/domjudge/blob/8d73e776dc9cdf1731d378d29373e5735d65b9c0/webapp/src/Service/SubmissionService.php#L534), which is the case for `.doc`. Uploading any other file, e.g. `.pdf` and selecting `C` as language will fail with the expected behavior described above. This has nothing to do with custom extensions and language selection, I think.

Of course, the user has to select the Language `C` manually, but the filter should step in (only if the option is activated). I agree that this should be expected behaviour.

The filter algorithm linked above is probably not perfect in this sense.

**Comment by HeRaNO** [comment]
> Hi @HeRaNO, I've tried to reproduce this and I can't. Can you maybe upload a bit more screenshots on how you submit this?
> 
> When I upload a file ending with the `.doc` syntax via submit the following field will stay empty: `Language` but for you this got picked up as `C`? Or did you manually select `C` as language?
> 
> If it's the first case this is indeed a bug, in the second case the message `match any of the allowed extensions` would be wrong as a participant can decide to hand something in with a custom extension (think about someone who can decide to use `.py2` vs `.py3`) and the system should just leave the language selection to the team IMO.

Hi and yes, it's the second case. The participant manually selects `C` language.

So the extension filter does not mean to `must submit files among these extensions`, it's only used to infer the language that participants use? Maybe I have some misunderstanding about it.

**Comment by ItsNiklas** [comment]
> > Hi @HeRaNO, I've tried to reproduce this and I can't. Can you maybe upload a bit more screenshots on how you submit this?
> > When I upload a file ending with the `.doc` syntax via submit the following field will stay empty: `Language` but for you this got picked up as `C`? Or did you manually select `C` as language?
> > If it's the first case this is indeed a bug, in the second case the message `match any of the allowed extensions` would be wrong as a participant can decide to hand something in with a custom extension (think about someone who can decide to use `.py2` vs `.py3`) and the system should just leave the language selection to the team IMO.
> 
> Hi and yes, it's the second case. The participant manually selects `C` language.
> 
> So the extension filter does not mean to `must submit files among these extensions`, it's only used to infer the language that participants use? Maybe I have some misunderstanding about it.

No, you are right: Enabling the filter should result in `must submit files among these extensions`. Language inference is always used.

There is just a small oversight in the code. I think I can work on fixing it.

**Comment by HeRaNO** [comment]
> > > Hi @HeRaNO, I've tried to reproduce this and I can't. Can you maybe upload a bit more screenshots on how you submit this?
> > > When I upload a file ending with the `.doc` syntax via submit the following field will stay empty: `Language` but for you this got picked up as `C`? Or did you manually select `C` as language?
> > > If it's the first case this is indeed a bug, in the second case the message `match any of the allowed extensions` would be wrong as a participant can decide to hand something in with a custom extension (think about someone who can decide to use `.py2` vs `.py3`) and the system should just leave the language selection to the team IMO.
> > 
> > 
> > Hi and yes, it's the second case. The participant manually selects `C` language.
> > So the extension filter does not mean to `must submit files among these extensions`, it's only used to infer the language that participants use? Maybe I have some misunderstanding about it.
> 
> No, you are right: Enabling the filter should result in `must submit files among these extensions`. Language inference is always used.
> 
> There is just a small oversight in the code. I think I can work on fixing it.

OK I got that. Thank you. 👍
